File: src/com/fictional_particle_sim/physicals/Field.java
package com.fictional_particle_sim.physicals;

import com.fictional_particle_sim.geometrics.BoundingBox;
import com.fictional_particle_sim.geometrics.Point;
import com.fictional_particle_sim.geometrics.Vector;

import java.awt.*;

import static com.fictional_particle_sim.Constants.*;

public class Field {

    public Point center, topLeft, bottomRight;
    public Vector force;
    public double radius, velocityScalar, chargeScalar;
    public String shape;
    public Color color;

    public BoundingBox boundingBox;

    public Field(Point topLeft, Point bottomRight, Color color, double velocityScalar, double chargeScalar) {
        this.topLeft = new Point(Math.min(topLeft.x, bottomRight.x), Math.min(topLeft.y, bottomRight.y));
        this.bottomRight = new Point(Math.max(topLeft.x, bottomRight.x), Math.max(topLeft.y, bottomRight.y));
        this.center = new Point();
        this.color = color;

        this.velocityScalar = ((velocityScalar - 1) * SPF) + 1;
        this.chargeScalar = ((chargeScalar - 1) * SPF) + 1;
        this.force = new Vector();

        shape = "RECTANGLE";
        this.boundingBox = boundingBox();
    }
    public Field(Point topLeft, Point bottomRight, Color color, Vector force) {
        this.topLeft = new Point(Math.min(topLeft.x, bottomRight.x), Math.min(topLeft.y, bottomRight.y));
        this.bottomRight = new Point(Math.max(topLeft.x, bottomRight.x), Math.max(topLeft.y, bottomRight.y));
        this.center = new Point();
        this.color = color;

        this.velocityScalar = 1;
        this.chargeScalar = 1;
        this.force = force;

        shape = "RECTANGLE";
        this.boundingBox = boundingBox();
    }

    public Field(Point center, double radius, Color color, double velocityScalar, double chargeScalar) {
        this.center = center;
        this.radius = radius;
        this.topLeft = new Point();
        this.bottomRight = new Point();
        this.color = color;

        this.velocityScalar = ((velocityScalar - 1) * SPF) + 1;
        this.chargeScalar = ((chargeScalar - 1) * SPF) + 1;
        this.force = new Vector();

        shape = "CIRCLE";
        this.boundingBox = boundingBox();
    }
    public Field(Point center, double radius, Color color, Vector force) {
        this.center = center;
        this.radius = radius;
        this.topLeft = new Point();
        this.bottomRight = new Point();
        this.color = color;

        this.velocityScalar = 1;
        this.chargeScalar = 1;
        this.force = force;

        shape = "CIRCLE";
        this.boundingBox = boundingBox();
    }

    public BoundingBox boundingBox() {

        return new BoundingBox(this);
    }
}

File: src/com/fictional_particle_sim/physicals/Field.kt
package com.fictional_particle_sim.physicals

import com.fictional_particle_sim.Main
import com.fictional_particle_sim.geometrics.BoundingBox
import com.fictional_particle_sim.geometrics.Point
import com.fictional_particle_sim.geometrics.Vector
import com.fictional_particle_sim.physicals.Shape.*
import com.fictional_particle_sim.util.scalarField
import com.fictional_particle_sim.util.vectorField
import java.awt.Color
import kotlin.math.pow

class Field(var topLeft: Point = Point(), var bottomRight: Point = Point(), var center: Point = Point(), var radius: Double = 0.0, var velocityScalar: scalarField = { 1.0 }, var chargeScalar: scalarField = { 1.0 }, var fieldForce: vectorField = {Vector()}, var shape: Shape, var color: Color) {

    fun boundingBox(): BoundingBox {
        var a = Point()
        var b = Point()
        when (shape) {
            CIRCLE -> {
                    a = center - Point(radius, radius)
                    b = center + Point(radius, radius)
                }
            RECTANGLE -> {
                    a = topLeft
                    b = bottomRight
                }
            else -> {}
        }
        return BoundingBox(a, b)
    }

    override fun toString(): String {
        return when (shape) {
            RECTANGLE -> "$shape, $topLeft, $bottomRight, ${scalarFieldToString(velocityScalar)}, ${scalarFieldToString(chargeScalar)}, ${vectorFieldToString(fieldForce)}"
            CIRCLE -> "$shape, $center, $radius, ${scalarFieldToString(velocityScalar)}, ${scalarFieldToString(chargeScalar)}, ${vectorFieldToString(fieldForce)}"
            else -> ""
        }
    }
}

fun attractiveField(center: Point, strength: Double): vectorField = {
    Vector(this, center).norm() * strength * (Vector(this, center).magnitude()).pow(2)
}

fun vectorFieldToString(fieldForce: vectorField): String {
    return "Pos -> Vel"
}
fun scalarFieldToString(fieldForce: scalarField): String {
    return "Pos -> Scalar"
}
